### PR TITLE
improvements to clustering

### PR DESCRIPTION
**Pull request by vhewes**
this feature branch overhauls how the instance decoder operates. the original scope of this work was to implement the adjusted Rand score (ARS) metric for clustering, but that work uncovered an issue with how we were performing clustering, so the scope has now expanded out to fixing that issue by redesigning pieces of the workflow

## the issue

after adding code to materialize predicted instance labels and calculate the ARS, i noticed that the model wasn't training well – and in fact, was returning nonsense values for that metric. the ARS metric ranges between -1 and 1 by construction, but we were logging values as extreme as 20. i dug into the code, and realized it was due to an oversight in how we construct the instance truth.

up until now, instance labels have been implemented as an integer tensor for each graph hit. any hits with a label of -1 are background – they do not belong to any instance. foreground hits have a label between 0 and N-1, where N is the number of true instances – so all hits associated with the first instance have a label of 0, all hits associated with the second instance have a label of 1, etc etc.

the oversight in how we batch was that integer labels were not offset between graphs in a batch – so every graph's instance labels started again at 0. since we were calculating object condensation across the batch, this meant that from the loss function's perspective, it looks like all hits from the first instance in every batch were all part of one large instance, and so on. this means the loss function would try to group together hits from completely separate events into the same instance, which obviously makes no sense.

## the solution

i considered two possible solutions:
1. offset instance labels across the batch, so every instance in every graph is distinct
2. calculate object condensation loss independently for each graph in the batch, and take the mean

### offsetting instance labels

PyG offers some simple solutions for offsetting tensors when batching graphs together, and one of those is the "increment dictionary." for each feature attribute, it stores a tensor of integers to add to each graph's tensor values when batching, and to subtract when unbatching. ideally we'd just use this functionality, but the fact that we use `-1` for the background hits complicates this – this mechanism would increment them too, which would make no sense.

instead, i added a transform to the graphs to remove the true instance label tensor on the hits, and instead encode the true instance labels as graph edges. each graph gets an empty node store `particle-truth`, with a number of nodes equal to the number of instances. we then assign the true instance labels as edges – each foreground hit is connected to the node corresponding to its true instance. if a hit doesn't belong to any instance in truth, then it doesn't get an edge. this means we don't have to assign any `-1` values, and so PyG automatically handles offsetting the instance indices across the batch.

when we come to compute the loss, we materialize those edges into a true instance label tensor on the hits. if the graphs are batched, the materialized true instance labels will be correctly offset across the batch – and if they aren't, it won't.

this seems to work really nicely – the model appears to be training very well, but it introduces issues with memory overhead. because we're offsetting instance labels, we now have many more true instance labels in a batch, which causes memory spikes due to the object condensation loss function, which is currently coded up to generate tensors internally with size `[N, K]` where `N` is the number of hits and `K` is the number of true instances. i plan to play around with the implementation of the object condensation loss, and see if we can avoid generating these large tensors. it will probably involve at least one for loop, which will slow things down, but if the alternative is out-of-memory errors, we can probably live with that in the short term.

### unbatched object condensation loss

i also tried unbatching the graphs and calculating the object condensation loss on each one independently, and then taking the mean. i was hoping this would be a better solution, but it didn't work at all. see the following instance loss curves:

Screenshot 2024-08-23 at 1 22 51 PM

here the orange is the batched object condensation curve with instance labels offset. it appears to be learning well, until it crashes due to an out-of-memory error, as discussed above. by contrast, the other two are with unbatched and average object condensation, trained by itself (green) and in tandem with the semantic and filter decoders (purple). neither of them trains well at all, so clearly the object condensation loss needs to be calculated once over the entire batch.

## other changes

i've also wrapped some other development changes into this branch:

### nugraph environment variables

it's always been irritating to switch between training on different clusters and need to constantly update paths to data files and log directories. i've started to build out a system which defines `NUGRAPH_DIR`, `NUGRAPH_DATA` and `NUGRAPH_LOG` environment variables to point towards standard directories on various hosts, so the code will run on different systems without requiring any modification. if the user exports these environment variables, then they will be used; if not, then at runtime the code will query the system's hostname and attempt to configure them automatically. this system needs to be finalized and propagated to all scripts and notebook before this pull request is merged.

### wandb improvements

i'm working to further improve the wandb logging infrastructure. this is reflected in changes to how and where we write output by default. it also means a change in how resuming an existing model training is implemented, but that part is not finished yet – so again, that will need to be finalized before this PR is merged.

### CODE REVIEW

**Comment by vhewes** [comment]
the instance-only training failed while i was typing the above comment 😭

**Comment by vhewes** [comment]
i started a new instance-only decoder training running using the same configuration as the 4-decoder training (batch size 16, 20 epochs) and it's running. we'll see how it performs in the end
